src/analyzer/protocol/tcp/TCP.cc

```diff
@@ -1209,8 +1218,28 @@ void TCP_Analyzer::DeliverPacket(int len, const u_char* data, bool is_orig,
 
 	// Handle child_packet analyzers.  Note: This happens *after* the
 	// packet has been processed and the TCP state updated.
-	LOOP_OVER_GIVEN_CHILDREN(i, packet_children)
-		(*i)->NextPacket(len, data, is_orig, rel_data_seq, ip, caplen);
+	analyzer_list::iterator next;
+
+	for ( auto i = packet_children.begin(); i != packet_children.end(); i = next )
+		{
+		auto child = *i;
+		next = ++i;
+
+		if ( child->IsFinished() || child->Removing() )
+			{
+			--i;
+
+			if ( child->Removing() )
+				child->Done();
+
+			DBG_LOG(DBG_ANALYZER, "%s deleted child %s",
+			        fmt_analyzer(this).c_str(), fmt_analyzer(child).c_str());
+			packet_children.erase(i);
```

Comment:
I don't think this is safe because erasing may invalidate the iterators. However, erase() returns an iterator pointing right after the deleted argument, so if we use that then, it should be well-defined. I'll make that change.

Reply:
For `std::list`: "References and iterators to the erased elements are invalidated", so think it's fine because `next` already advanced and we have no other usages of `i` after erasing it ?

(I was also mostly just copying the pattern already used several places for `Analyzer::DeleteChild` and its usages)

Reply:
Ah, for some reason I thought it was a vector. Yes, for lists I think the code was fine. Anyways, the new one should work in either case.